edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/authenticators/DeviceScopeTokenAuthenticator.cs

```diff
@@ -0,0 +1,302 @@
+            this.edgeHubHostName = Preconditions.CheckNonWhiteSpace(edgeHubHostName, nameof(edgeHubHostName));
+            this.connectionManager = Preconditions.CheckNotNull(connectionManager, nameof(connectionManager));
+            this.deviceScopeIdentitiesCache.ServiceIdentityUpdated += this.HandleServiceIdentityUpdate;
+            this.deviceScopeIdentitiesCache.ServiceIdentityRemoved += this.HandleServiceIdentityRemove;
+        }
+
+        async void HandleServiceIdentityUpdate(object sender, ServiceIdentity serviceIdentity)
+        {
+            try
+            {
+                Events.ServiceIdentityUpdated(serviceIdentity.Id);
+                Option<IClientCredentials> clientCredentials = this.connectionManager.GetClientCredentials(serviceIdentity.Id);
+                await clientCredentials.ForEachAsync(
+                    async c =>
+                    {
+                        if (!(c is ITokenCredentials tokenCredentials) ||
+                            !await this.AuthenticateInternalAsync(tokenCredentials, serviceIdentity))
+                        {
+                            Events.ServiceIdentityUpdatedRemoving(serviceIdentity.Id);
+                            await this.connectionManager.RemoveDeviceConnection(c.Identity.Id);
+                        }
+                        else
+                        {
+                            Events.ServiceIdentityUpdatedValidated(serviceIdentity.Id);
+                        }
+                    });
+            }
+            catch (Exception ex)
+            {
+                Events.ErrorReauthenticating(ex, serviceIdentity);
+            }
+        }
+
+        async void HandleServiceIdentityRemove(object sender, string id)
+        {
+            try
+            {
+                Events.ServiceIdentityRemoved(id);
+                await this.connectionManager.RemoveDeviceConnection(id);
+            }
+            catch (Exception ex)
+            {
+                Events.ErrorRemovingConnection(ex, id);
+            }
+        }
+
+        public async Task<bool> AuthenticateAsync(IClientCredentials clientCredentials)
+        {
+            if (!(clientCredentials is ITokenCredentials tokenCredentials))
+            {
+                return false;
+            }
+
+            try
+            {
+                Option<ServiceIdentity> serviceIdentity = await this.deviceScopeIdentitiesCache.GetServiceIdentity(clientCredentials.Identity.Id);
+                return await serviceIdentity.Map(s => this.AuthenticateInternalAsync(tokenCredentials, s))
+                    .GetOrElse(() => this.underlyingAuthenticator.AuthenticateAsync(clientCredentials));
+            }
+            catch (Exception e)
+            {
+                Events.ErrorAuthenticating(e, clientCredentials);
+                return await this.underlyingAuthenticator.AuthenticateAsync(clientCredentials);
```

Comment:
>return await this.underlyingAuthenticator.AuthenticateAsync(clientCredentials); [](start = 16, length = 79)

Should this be called only when AuthenticateInternalAsync throws? Or if underlyingAuthenticator fails does it make sense to call it again?